First Draft Heating Boost

albarlow opened the pull request:
Submitted for review, testing and comments - I'm not hugely familiar with all the elements of integrations but I had a look at how some other integrations handle service calls and boosts as well as the API documentation and python library that underpins this integration and have added in a service call that will target a neomiser climate entity and apply a boost for a set amount of hours, minutes and degrees.  All the values have default fallbacks, I defaulted to 20deg, 1 hr, 0 mins.  This allows for a 'default boost' of an hour if you specify no parameters.  I incorporated a sanity check on the times requested as these can cause the request to fail.

I don't know if there is a way to make the services pull through to the device as at the moment you have to go looking for them, but I don't know if that's just the way it is with all custom services.

## Set Hold (Boost On) Service Call with Fields
![image](https://user-images.githubusercontent.com/60830942/209034923-40d2d524-a182-4ad0-b3be-a0fdf0f3b37b.png)
## Set Hold (Boost On) Service Call with Yaml
![image](https://user-images.githubusercontent.com/60830942/209034958-51840e54-52ad-422e-ae6c-d8e8b94f9011.png)
## Cancel Hold (Boost Off) Service Call with Fields (no parameters)
![image](https://user-images.githubusercontent.com/60830942/209035002-d2ecaac4-ea79-42ce-941d-581cf0fd946f.png)

I don't know if there's any legs for doing something like WLED where they have sliders to control settings on the device. 
![image](https://user-images.githubusercontent.com/60830942/209034176-9901535a-9002-49e4-9989-9d06a0163c8f.png)

Lastly, I have wavered around whether to use the heatmiser 'Hold' terminology or the more common parlence 'Boost'.  For the attributes I've used 'Hold' as this is pulled from the device, but for the service call, I used Boost as that's perhaps more self explanatory?  I don't have any cooling or heating/cooling devices to test this on and have no idea if it will work for either of those use cases. 

Ref also my recent PR #118 and Issue #9

Trouble-51 commented:
Hi,

 

Shouldn’t the fields 

 

Boost minutes

Boost temperature and 

Boost hours (combine this with boost minutes (see below)

 

And add a list of the various climate.* 

 

have drop-down menus behind which give you an input select with the options like

 

  set_temperature:

    name: Hold Temperature At

    options:

      - '19'

      - '20'

      - '21'

      - '22'

      - '23'

      - '24'

      - '25'

      - '26'

      - '27'

    initial: '21'

    icon: mdi:target

 

  set_time:

    name: Set Time

    options:

      - 30 minutes

      - 1 hour

      - 2 hours

      - 3 hours

      - 4 hours

    initial: 30 minutes

    icon: mdi:clock

 

  set_room:

    name: Select Room

    options:

      - 'Study'

      - 'Living & Kitchen'

      - 'Master Bedroom'

      - 'Master Bathroom'

      - 'Toilet Upstairs'

      - 'Hall Downstairs'

      - 'Bedroom Nicoline'

      - 'Bedroom Jeroen'

      - 'Bedroom Floor'

      - 'Bedroom M & Y'

      - 'Bedroom Kleinkinderen'

      - 'Bathroom N & J'

      - 'Bathroom M & Y'

    initial: 'Living & Kitchen'

    icon: mdi:home

 

 

 

which would look something like this:

 

 

 

 

This is something I made which includes in the config/configuration file:

 

  - platform: template

    sensors:

      set_room:

        friendly_name: Set Room

        value_template:

          "{% if is_state('input_select.set_room' , 'Study') %} climate.study

          {% elif is_state('input_select.set_room' , 'Living & Kitchen') %} climate.living_and_kitchen

          {% elif is_state('input_select.set_room' , 'Master Bedroom') %} climate.master_bedroom_2

          {% elif is_state('input_select.set_room' , 'Master Bathroom') %} climate.master_bathroom

          {% elif is_state('input_select.set_room' , 'Toilet Upstairs') %} climate.toilet_upstairs

          {% elif is_state('input_select.set_room' , 'Hall Downstairs') %} climate.hall_downstairs

          {% elif is_state('input_select.set_room' , 'Bedroom Nicoline') %} climate.nicoline

          {% elif is_state('input_select.set_room' , 'Bedroom Jeroen') %} climate.jerpen

          {% elif is_state('input_select.set_room' , 'Bedroom Floor') %} climate.bedroom_floor

          {% elif is_state('input_select.set_room' , 'Bedroom M & Y') %} climate.mark_and__ymke

          {% elif is_state('input_select.set_room' , 'Bedroom Kleinkinderen') %} climate.kleinkinderen

          {% elif is_state('input_select.set_room' , 'Bathroom N & J') %} climate.bathroom_nicoline_jeroen

          {% elif is_state('input_select.set_room' , 'Bathroom Mark & Ymke') %} climate.y_and_y_bathroom

          {% endif %}"

 

  - platform: template

    sensors:

      set_temperature:

        friendly_name: Set Temperature

        value_template:

          "{% if is_state('input_select.set_temperature' , '19') %}} 19

          {% elif is_state('input_select.set_temperature' , '20') %} 20

          {% elif is_state('input_select.set_temperature' , '21') %} 21

          {% elif is_state('input_select.set_temperature' , '22') %} 22

          {% elif is_state('input_select.set_temperature' , '23') %} 23

          {% elif is_state('input_select.set_temperature' , '24') %} 24

          {% elif is_state('input_select.set_temperature' , '25') %} 25

          {% elif is_state('input_select.set_temperature' , '26') %} 26

          {% elif is_state('input_select.set_temperature' , '27') %} 27

          {% endif %}"

 

  - platform: template

    sensors:

      start_timer:

        friendly_name: Start Timer

        value_template:

          "{% if is_state('input_select.set_time' , '30 minutes') %} 00:30:00

          {% elif is_state('input_select.set_time' , '1 hour') %} 01:00:00

          {% elif is_state('input_select.set_time' , '2 hours') %} 02:00:00

          {% elif is_state('input_select.set_time' , '3 hours') %} 03:00:00

          {% elif is_state('input_select.set_time' , '4 hours') %} 04:00:00

          {% endif %}"

 

Which wouldn’t be needed if the boost function (I called it Set & Hold Temperature) is part of the Heatmiser integration

 

regards

 

From: albarlow ***@***.***>
Reply to: MindrustUK/Heatmiser-for-home-assistant ***@***.***>
Date: Thursday, 22 December 2022 at 01:30
To: MindrustUK/Heatmiser-for-home-assistant ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [MindrustUK/Heatmiser-for-home-assistant] First Draft Heating Boost (PR #119)

 

Submitted for review, testing and comments - I'm not hugely familiar with all the elements of integrations but I had a look at how some other integrations handle service calls and boosts as well as the API documentation and python library that underpins this integration and have added in a service call that will target a neomiser climate entity and apply a boost for a set amount of hours, minutes and degrees. All the values have default fallbacks, I defaulted to 20deg, 1 hr, 0 mins. This allows for a 'default boost' of an hour if you specify no parameters. I incorporated a sanity check on the times requested as these can cause the request to fail.

I don't know if there is a way to make the services pull through to the device as at the moment you have to go looking for them, but I don't know if that's just the way it is with all custom services.
Set Hold (Boost On) Service Call with Fields
Set Hold (Boost On) Service Call with Yaml
Cancel Hold (Boost Off) Service Call with Fields (no parameters)
I don't know if there's any legs for doing something like WLED where they have sliders to control settings on the device.

Lastly, I have wavered around whether to use the heatmiser 'Hold' terminology or the more common parlence 'Boost'. For the attributes I've used 'Hold' as this is pulled from the device, but for the service call, I used Boost as that's perhaps more self explanatory? I don't have any cooling or heating/cooling devices to test this on and have no idea if it will work for either of those use cases.

Ref also my recent PR #118 and Issue #9

You can view, comment on, or merge this pull request online at:
  https://github.com/MindrustUK/Heatmiser-for-home-assistant/pull/119
Commit Summary
2f6ebb4 Initial Hold Functionality
3e32399 Sense-check Times
File Changes 
(3 files)
M custom_components/heatmiserneo/climate.py (68) 
M custom_components/heatmiserneo/const.py (6) 
A custom_components/heatmiserneo/services.yaml (43) 
Patch Links:
https://github.com/MindrustUK/Heatmiser-for-home-assistant/pull/119.patch
https://github.com/MindrustUK/Heatmiser-for-home-assistant/pull/119.diff
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

Trouble-51 commented:
This is what I get when I include the changes to the files:

 

 

But that doesn’t allow me to input the data time, temperature and which climate*room  

 

 

 

From: albarlow ***@***.***>
Reply to: MindrustUK/Heatmiser-for-home-assistant ***@***.***>
Date: Thursday, 22 December 2022 at 01:30
To: MindrustUK/Heatmiser-for-home-assistant ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [MindrustUK/Heatmiser-for-home-assistant] First Draft Heating Boost (PR #119)

 

Submitted for review, testing and comments - I'm not hugely familiar with all the elements of integrations but I had a look at how some other integrations handle service calls and boosts as well as the API documentation and python library that underpins this integration and have added in a service call that will target a neomiser climate entity and apply a boost for a set amount of hours, minutes and degrees. All the values have default fallbacks, I defaulted to 20deg, 1 hr, 0 mins. This allows for a 'default boost' of an hour if you specify no parameters. I incorporated a sanity check on the times requested as these can cause the request to fail.

I don't know if there is a way to make the services pull through to the device as at the moment you have to go looking for them, but I don't know if that's just the way it is with all custom services.
Set Hold (Boost On) Service Call with Fields
Set Hold (Boost On) Service Call with Yaml
Cancel Hold (Boost Off) Service Call with Fields (no parameters)
I don't know if there's any legs for doing something like WLED where they have sliders to control settings on the device.

Lastly, I have wavered around whether to use the heatmiser 'Hold' terminology or the more common parlence 'Boost'. For the attributes I've used 'Hold' as this is pulled from the device, but for the service call, I used Boost as that's perhaps more self explanatory? I don't have any cooling or heating/cooling devices to test this on and have no idea if it will work for either of those use cases.

Ref also my recent PR #118 and Issue #9

You can view, comment on, or merge this pull request online at:
  https://github.com/MindrustUK/Heatmiser-for-home-assistant/pull/119
Commit Summary
2f6ebb4 Initial Hold Functionality
3e32399 Sense-check Times
File Changes 
(3 files)
M custom_components/heatmiserneo/climate.py (68) 
M custom_components/heatmiserneo/const.py (6) 
A custom_components/heatmiserneo/services.yaml (43) 
Patch Links:
https://github.com/MindrustUK/Heatmiser-for-home-assistant/pull/119.patch
https://github.com/MindrustUK/Heatmiser-for-home-assistant/pull/119.diff
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

albarlow commented:
I could well be wrong here, but I think you would have to keep your own entity selector to achieve what you're after...I've not found any dashboard card that presents the nice call-service interface from Developer Tools/Automation engine for the end-user.

There is nothing to stop you using your prepopulated fields to call the service I've drafted.  The way that custom services seem to work is that there is a yaml file that contains the variables and associated fields, similar to how you define an automation or script and then you have code within the climate entity that picks up the call.  Your example, as I understand it, would require the user to populate and maintain a curated list with translations, it must be possible to do, as I've seen it in other parts of HA, but I've not seen a way to pull it into the dashboards.

With the time selectors, I didn't go for a list of pre-defined times in order to preserve the full functionality of the holds function.  I have separate hours and minutes as fields as that is how it is passed to the stat in the API call (see below).  You could potentially make it a duration selector and parse it into hours and minutes.
![image](https://user-images.githubusercontent.com/60830942/209196310-81ebf4d8-0d0a-4f6b-ba7a-6a9ec37ff52b.png)

As it stands, you have a selector that aliases times that align to your preferences for time slots, which is great and clearly works for you, but how do you account for the breadth of users who might want all sorts of durations?  Maybe some people want 30 minute granularity, others 15 minute, others to the minute.  You could still use your list and map that to the service call.  

We could constrain the minutes to being intervals of 5 as the chances that somebody wants precisely 31 minutes is unlikely :)

The main reason I wrote this was that without it, the only way to set a hold/boost other than their app that I could find was to write a whole NodeRED flow (per my examples in #118 ).  Now you can operate holds fully within HA.

If you did want to have a play with the function, this is the one (near the bottom of climate.py) to look at.
![image](https://user-images.githubusercontent.com/60830942/209215600-3226e633-a0c2-4bba-8362-be527322fc4a.png)

...and then adjust accordingly in services.yaml in the same folder.

albarlow commented:
Made a commit that tries out the duration selector.  I'm not getting the selector show up in the GUI, not sure why.  I don't know if Home Assistant has pre-written script the handles the different formats of duration but I made my own little script to handle it for now.

For some reason I can't get the selector to appear in the GUI...it did before but now it doesn't.

You can call it in yaml though passing either boost_duration: '01:00', boost_duration: '01:00:00' or boost_duration: hours: 1 minutes: 0

Trouble-51 commented:
Don’t you need a (python)script to put the attribute ‘hold_on’ to ‘true’ as otherwise the sensors will (I think) always be in state ‘false’, before you run the boost_heating script?

 

The selectors would work if they were on different lines as opposed to multiple-entity-row and you could then set them in the frontend as opposed to in the script. And the script would then apply the settings to the Target.  I haven’t found a way to get the drop-down menus in the multiple-entity-row version, which would be much nicer though. Running the current script doesn’t give me the input options. 

 

In the Boost column I would put the remaining time, not the total time of the boost.

 

Just some thoughts 

 

From: albarlow ***@***.***>
Reply to: MindrustUK/Heatmiser-for-home-assistant ***@***.***>
Date: Saturday, 24 December 2022 at 21:19
To: MindrustUK/Heatmiser-for-home-assistant ***@***.***>
Cc: Trouble-51 ***@***.***>, Comment ***@***.***>
Subject: Re: [MindrustUK/Heatmiser-for-home-assistant] First Draft Heating Boost (PR #119)

 

Made a commit that tries out the duration selector. I'm not getting the selector show up in the GUI, not sure why. I don't know if Home Assistant has pre-written script the handles the different formats of duration but I made my own little script to handle it for now.

For some reason I can't get the selector to appear in the GUI...it did before but now it doesn't.

You can call it in yaml though passing either boost_duration: '01:00', boost_duration: '01:00:00' or boost_duration: hours: 1 minutes: 0

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you commented.Message ID: ***@***.***>

albarlow commented:
Hi, I think we might be at cross-purposes regarding what the pull request implements and I've probably muddied the waters by putting two separate PRs in on similar topics at the same time...

Adding the hold_on, hold_temp and hold_time attribute to the climate entity was just outputting to the user something the underlying python library was already getting from the hub.  This just made it part of the climate entity data.  I used Home Assistant templating to make the additional template entities based on the attribute and Helpers/Custom Cards to make the selector for the boost.

_This_ PR adds a service that is attached to the integration itself and lets you target a specific heatmiser climate entity, or all the heatimiser climate entities in an area (e.g. Downstairs) and have them each request to boost themselves.  This means each climate entity sends a boost request via the hub for their individual stat.  Some period later, the main integration polls the devices to get their updated status.  This is why you don't get instant feedback to give you confidence that your call worked and that's why the boost script doesn't need to toggle the attribute statues.

That did prompt me to notice that when you set temperature, it 'optimistically' assumes at that the command worked.  I have applied this same logic to starting or ending a boost.  It immediately sets the entity attributes as if the change had already happened.  If it fails to apply, the next poll cycle will set it back to what it actually is, but if it works, it provides that snappier response we expect to see.

Regarding showing the the boost time, the example I posted does show the remaining boost time and that attribute is drawn straight from the API.  With regard to adding time versus setting the new remaining time, you could feasibly write a service to do that, read the existing time remaining attribute and adding the time on but for me at least, I don't fancy the extra work to do that :)

I still don't know why the duration box doesn't render on the developer>services page.
I don't know if home assistant has a built in decoder for durations that would read the duration more neatly.

PTomlinson1 commented:
Great news that you're working on the Hold function - I have a heatmiser system and it's the main way I heat rooms for occupancy and have really been wanting to use that from HA.

If it helps, a Heatmiser NeoStat user will be used to the Heatmiser app. This means it makes sense to refer to the Hold function as the Hold function, rather than using the term Boost.  

In the Heatmiser app, when using the hold function, the drop down lets you change temperature in increments of 0.5deg. For the hold time, it starts in increments of 30 minutes up to 3 hours, then increments of 1 hour up to 8 hours.  On the thermostat itself you can adjust the hold time in increments of 30 minutes up to 99 hours.

So whilst I totally understand that there is the flexibility in HA to allow selection of very precise hold periods, any Heatmiser user is already used to chunks of 30 minutes. 

So maybe allow the service to be called with any time that is passed to it (up to 99 hours) but if you're designing a selector then  you can probably keep it simple and stick to chunks of 30 minutes - similar to the drop down lists in one of the posts above.

I'll see if I can play with your code and test it.

Again, many thanks for this really useful work.

Trouble-51 commented:
I think that the HA convention for time is expressed as '00:00:00'  hours, minutes, seconds or '00:00' just hours and minutes. Maybe that why the duration doesn't work?

albarlow commented:
@PTomlinson1 thanks for that - I don't think you can do the 30min increments in the backend but you could do that in the frontend through a dashboard card but doing it this way would definitely allow for that.  I think you make a good point about people being familiar with the code already needing to be familiar with the Hold nomenclature.  I've just gone through and renamed all the references the Boosts into references to Holds.

I've also added to the readme.md about holds.

@Trouble-51 It seems to depend on how you call the service - if you call it with `duration: '1:15' or '1:15:00', it just passes that string through; however, if you call the variable like this
```yaml
duration:
    hours: 1
    minutes: 15
```
...it comes out as an objet along the lines of `{duration:{hours:1,minutes:15}}`

Trouble-51 commented:
Hi, 

Happy new Year

Can you not use the drop down option in helpers? e.g. for time:

<img width="526" alt="Screenshot 2023-01-02 at 11 19 28" src="https://user-images.githubusercontent.com/26715092/210224563-373290ec-6e5d-4b72-86f2-b55a2981479a.png">

and the same for the various thermostats and temperature?

PTomlinson1 commented:
I've had a play and the Hold function works well.

I've tried creating a couple of drop downs for time and temperature and that can work.

The real challenge is fitting them nicely on a card, especially one that will fit on a phone screen.

I have no idea how to create a custom card, but it strikes me that is what is needed to make this "perfect" - a card with the thermostat name, status (hvac_action), current temperature,  current set temp,  + / - buttons to choose Hold time, + / - buttons to choose Hold temp, and a button to call the function with the chosen parameters. Maybe also a display of Hold time  remaining if already set.

In the meantime just the ability to send the Hold command is a massive bonus.. Thanks for all your work.

albarlow commented:
The most concise/compact option I've found with in a single card is with the Mini Climate Card (available on HACS).  Only thing that doesn't seem to work is that it should be able to turn the boost button to active color when hold_on is true but for some reason it won't evaluate it.  I've not ben able to figure out why.  You can also add in things like humidity sensors, fans etc.

```yaml
template: mini-climate-card
template_data:
  zone: kitchen
type: custom:mini-climate
entity: climate.kitchen
name: Thermostat
icon: mdi:heat-wave
toggle:
  default: 'on'
secondary_info:
  type: hvac-action
  source:
    idle:
      icon: mdi:snowflake
      name: Idle
    heating:
      icon: mdi:fire
      name: Heating
hvac_mode:
  source:
    'off':
      icon: mdi:thermostat
      name: Standby Mode
    heat:
      icon: mdi:thermostat-auto
      name: Timer Mode
fan_mode:
  hide: true
indicators:
  hold_on:
    icon: mdi:fire
    source:
      entity: climate.kitchen
      attribute: hold_on
      values:
        'true': Boost On
        'false': Boost Off
buttons:
  hold:
    type: dropdown
    name: Boost
    icon: mdi:fire-circle
    style: >
      (entity) => (entity.entity_id == 'true' ? {color:
      '--paper-item-icon-active-color'} : {})
    state:
      attribute: hold_on
    source:
      '0': 'Off'
      '30': 30 mins
      '60': 1 hour
      '90': 1 hour 30 mins
      '120': 2 hours
    change_action: >
      (selected_value, state, entity) => this.call_service('heatmiserneo',
      'hold_on', { entity_id: entity.entity_id, hold_duration: {hours: 0,
      minutes: parseInt(selected_value)}})
```

This works because I wrote it to parse any amount of minutes over 60 into hours and minutes.  You could add more values or change the values to your preferred boosts.

PTomlinson1 commented:
Have had a play with that - what temperature is being used for the hold?  Seems that the temperature picker changes the normal set temperature for the current schedule, and then also uses it as the hold temperature after picking the hold time from the dropdown.   So when you cancel the hold it's still set at the same set temperature until the end of the current scheduled period.   Don't know if you can stop it changing the set temp. Will keep playing.

PTomlinson1 commented:
Played a bit more, and found that if I call the service from the dropdown or the temperature selector, and  I call the service with only one of the 2 parameters (hold temp or duration) then it is using the default value for the missing parameter, which appears to be defined in services.yaml. (1 hour duration, 20 C)

Is it possible to make the call with only one parameter, so that the other is unchanged?

E.g. let's say hold was set 20 minutes ago at 21c for 2 hours. I want to reduce the hold temp to 20, but not change the remaining duration,.

another e.g. hold is not set, but I want to change the hold temp (e.g. changing the selector on the card) before then choosing the duration.

Thanks  again for your help - I might be making progress with how to use that card, but stuck because of spotting this behaviour.

PTomlinson1 commented:
Thought I'd post my card code in case it helps - very work in progress, but I was thinking you could use this just for the hold function, and another version for the normal set temp / current temp.
`type: custom:mini-climate
entity: climate.study_2
name: Thermostat Hold
icon: mdi:radiator
toggle:
  hide: true
hide_current_temperature: true
target_temperature:
  icons:
    up: mdi:chevron-up
    down: mdi:chevron-down
  unit: °C
  min: 14
  max: 26
  step: 0.5
  source:
    attribute: hold_temp
  change_action: >
    (value, entity) => this.call_service('heatmiserneo', 'hold_on', { entity_id:
    entity.entity_id, hold_temperature: value })
secondary_info:
  type: hvac-action
  source:
    idle:
      icon: mdi:snowflake
      name: Idle
    heating:
      icon: mdi:fire
      name: Heating
fan_mode:
  hide: true
hvac_mode:
  hide: true
indicators:
  hold_on:
    icon: mdi:clock
    source:
      entity: climate.study_2
      attribute: hold_time
buttons:
  hold:
    type: dropdown
    name: Hold
    icon: mdi:timer
    location: main
    state:
      attribute: hold_on
    style: >
      (entity) => (entity.entity_id == 'true' ? {color:
      '--paper-item-icon-active-color'} : {})
    source:
      '0': 'Off'
      '30': 30 mins
      '60': 1 hour
      '90': 1 hour 30 mins
      '120': 2 hours
    change_action: >
      (selected_value, state, entity) => this.call_service('heatmiserneo',
      'hold_on', { entity_id: entity.entity_id, hold_duration: {hours: 0,
      minutes: parseInt(selected_value)}})
`

PTomlinson1 commented:
Oops - how do you post YAML correctly?

albarlow commented:
That's great, @MindrustUK - hopefully we can get the PR merged.

I don't know if some other changes have come in since I opened the PR as it's now got some conflicts.  The readme file looks straightforward to resolve but I don't know exactly what changes have been made to climate.py recently.

MindrustUK commented:
I started working through the conflicts, I'm hoping to get this merged then I can resume a few other bits.

albarlow commented:
Brilliant, thanks so much!  In my testing on my system at home it seemed to be stable.  That was based on upstream as it was around December, I haven't touched it in the last few months, so hopefully it should come together ok.

MindrustUK commented:
@albarlow  I've merged this into branch: beta_testing-hold please pull this and give it a test. Everything seems to be working as expected. If you can feedback then I can pull this into master.

albarlow commented:
@MindrustUK I've given this a test on my devices this week - the hold functionality seems to work as intended.  Seems to me that we're good to go.

MindrustUK commented:
@albarlow I've merged this to master, I've not seen any issues in my testing. Thanks very much for your contribution to the project!
